src/main/java/org/gaul/s3proxy/EncryptedBlobStore.java

```diff
@@ -0,0 +1,176 @@
+
+        String peacemakrApiKey = properties.getProperty(S3ProxyConstants.PROPERTY_PEACEMAKR_API_KEY);
+        String clientName = properties.getProperty(S3ProxyConstants.PROPERTY_CLIENT_NAME);
+        if (clientName.isEmpty()) {
+            clientName = "s3proxy" + properties.getProperty(S3ProxyConstants.PROPERTY_IDENTITY);
+        }
+
+        try {
+            peacemakrSDK = Factory.getCryptoSDK(peacemakrApiKey, clientName, null, new InMemoryPersister(), null);
+            peacemakrSDK.register();
+        } catch (PeacemakrException e) {
+            Throwables.propagate(e);
+        }
+    }
+
+    static BlobStore newEncryptedBlobStore(BlobStore blobStore,
+                                           Properties properties) {
+        return new EncryptedBlobStore(blobStore, properties);
+    }
+
+    private byte[] getByteArrayFromStream(Blob blob) throws IOException {
+        InputStream blobStream = blob.getPayload().openStream();
+        ByteArrayOutputStream os = new ByteArrayOutputStream();
+
+        byte[] buffer = new byte[1024];
+        int len;
+
+        // read bytes from the input stream and store them in buffer
+        while ((len = blobStream.read(buffer)) != -1) {
+            // write bytes from the buffer into output stream
+            os.write(buffer, 0, len);
+        }
+
+        return os.toByteArray();
+    }
+
+    private Blob rebuildWithNewPayload(String container, Blob blob, byte[] newPayload) {
+        BlobMetadata blobMeta = blob.getMetadata();
+        ContentMetadata contentMeta = blob.getMetadata().getContentMetadata();
+        Blob newBlob = blobBuilder(container)
+                .name(blobMeta.getName())
+                .type(blobMeta.getType())
+                .tier(blobMeta.getTier())
+                .userMetadata(blobMeta.getUserMetadata())
+                .payload(newPayload)
+                .cacheControl(contentMeta.getCacheControl())
+                .contentDisposition(contentMeta.getContentDisposition())
+                .contentEncoding(contentMeta.getContentEncoding())
+                .contentLanguage(contentMeta.getContentLanguage())
+                .contentLength(newPayload.length)
+                .contentType(contentMeta.getContentType())
+                .contentMD5((HashCode) null)
+                .build();
+
+        newBlob.getMetadata().setUri(blobMeta.getUri());
+        newBlob.getMetadata().setETag(blobMeta.getETag());
+        newBlob.getMetadata().setLastModified(blobMeta.getLastModified());
+        newBlob.getMetadata().setSize(blobMeta.getSize());
+        newBlob.getMetadata().setPublicUri(blobMeta.getPublicUri());
+        newBlob.getMetadata().setContainer(blobMeta.getContainer());
+
+        return newBlob;
+    }
+
+    private Blob encryptBlob(String container, Blob blob) {
+        try {
+            byte[] payload = getByteArrayFromStream(blob);
+            byte[] encrypted = peacemakrSDK.encryptInDomain(payload, "works2");
```

Comment:
Maybe we should use "default" domain or mention this in the README? :)

Reply:
Agree :)